src/lib/shell/MainLoopZephyr.cpp

```diff
@@ -14,30 +14,116 @@
+ * @param help                    The help text for the command at roote node of the tree. Set to NULL if none.
+ *
+ * @return                        The shell_static_entry structure tree that chains all the sub-commands from the
+ *                                requested command node.
+ *
+ */
+const struct shell_static_entry * build_command_tree(const char * prefix, const char * syntax, const char * help)
+{
+
+    std::string sub_prefix_str = prefix;
+
+    // concatenate prefix and syntax to construct the prefix string for command suggestion.
+
+    if (strcmp(prefix, "") != 0)
+    {
+        sub_prefix_str += " ";
+    }
+
+    sub_prefix_str += syntax;
+
+    const char * sub_prefix = sub_prefix_str.c_str();
+
+    std::vector<shell_command_t *> subcmds = Engine::GetCommandSuggestions(sub_prefix);
+
+    const shell_static_entry * static_entry;
+
+    if (subcmds.size() > 0)
+    {
+        // Non-leaf node, recursively build tree for each child node
+        const struct shell_static_entry * sub_static_entry = new shell_static_entry[subcmds.size() + 1];
+        for (size_t i = 0; i < subcmds.size(); i++)
+        {
+            const struct shell_static_entry * sub_entry =
+                build_command_tree(sub_prefix, subcmds[i]->cmd_name, subcmds[i]->cmd_help);
+            memcpy((shell_static_entry *) &sub_static_entry[i], sub_entry, sizeof(struct shell_static_entry));
+        }
+
+        // Zephyr's shell_cmd_entry.u.entry needs an additional NULL element to terminate the array.
+        memset((shell_static_entry *) &sub_static_entry[subcmds.size()], 0, sizeof(struct shell_static_entry));
+
+        const struct shell_cmd_entry * sub_cmd_entry =
+            new const shell_cmd_entry{ .is_dynamic = false,
+                                       .u          = { .entry = (const struct shell_static_entry *) sub_static_entry } };
+
+        // The non-leaf nodes should require at least 1 required argument (for its child)
+        // Set the root command to use syntax "matter"
+        // Only the root "matter" node should have a handler function in Zephry-shell, so that all commands are sent
+        // to be dispatched by their registered Engine instances.
+        static_entry = new const shell_static_entry{ .syntax  = (strcmp(syntax, "") == 0) ? "matter" : syntax,
+                                                     .help    = help,
+                                                     .subcmd  = (const struct shell_cmd_entry *) sub_cmd_entry,
+                                                     .handler = (strcmp(syntax, "") == 0) ? cmd_matter : NULL,
+                                                     .args    = { 1, 10 } };
+    }
+    else
+    {
+        // the leaf nodes should have 0 required arguments but accepts a few (set 10 here) optional arguments.
+        static_entry =
+            new const shell_static_entry{ .syntax = syntax, .help = help, .subcmd = NULL, .handler = NULL, .args = { 0, 10 } };
```

Comment:
I think this should not return a pointer if it's then copied to a static variable (and the pointer is leaked).

Reply:
thanks you're right. I updated the code and do you mind taking another look? Thanks!